index.js

```diff
@@ -36,88 +36,187 @@ module.exports = function(options) {
-      if (file.contents.toString() !== file.originalContent) {
-        next();
-        return;
-      }
+  file.basename = file.basename.replace(/^_/, '.');
+  file.basename = file.basename.replace(/^\$/, '');
+  next(null, file);
+}
 
-      file.contents = new Buffer(JSON.stringify(file.json, null, 2) + '\n');
-      next();
-    });
+/**
+ * Uses C-style macros to escape templates with `{%%= foo %}` or
+ * `<%%= foo %>` syntax, so they will not be evaluated by a template
+ * engine when `.render` is called.
+ *
+ * @name escape templates
+ * @api public
+ */
 
-    return plugin;
-  };
-};
+function escape(file, next) {
+  if (file.isNull() || file.isBinary()) {
+    next(null, file);
+    return;
+  }
+
+  var str = file.contents.toString();
+  str = str.replace(/([{<])(%%[=-]?)/g, '__ESC_$1DELIM$2__');
+  file.contents = new Buffer(str);
+  next(null, file);
+}
+
+/**
+ * Expose `unescape`
+ */
 
 function unescape(file, next) {
+  if (file.isNull() || file.isBinary()) {
+    next(null, file);
+    return;
+  }
+
   var str = file.contents.toString();
-  str = str.replace(/__ESC_([{<])DELIM%(%=?)__/g, '$1$2');
+  str = str.replace(/__ESC_([{<])DELIM%(%[=-]?)__/g, '$1$2');
   file.contents = new Buffer(str);
   next();
 }
 
 /**
- * Expose `unescape`
+ * Adds a `json` property to the `file` object when the file extension
+ * matches `options.jsonRegex`. This allows JSON files to be updated
+ * by other middleware or pipeline plugins without having to parse and
+ * stringify with each modification.
+ *
+ * @name JSON on-load
+ * @api public
  */
 
-module.exports.unescape = unescape;
+function parseJson(file, next) {
+  if (file.extname !== '.json') {
+    next(null, file);
+    return;
+  }
+
+  var str = file.contents.toString();
+  utils.define(file, 'originalContent', str);
+  var json;
+
+  Object.defineProperty(file, 'json', {
+    configurable: true,
+    enumerable: true,
+    set: function(val) {
+      json = val;
+    },
+    get: function() {
+      return json || (json = JSON.parse(str));
+    }
+  });
+
+  next(null, file);
+}
+
+/**
+ * If `file.contents` has not already been updated directly, the `file.contents` property
+ * is updated with stringified JSON before writing the file back to the file
+ * system.
+ *
+ * @name JSON pre-write
+ * @api public
+ */
+
+function updateJson(file, next) {
+  if (file.contents.toString() !== file.originalContent) {
+    next();
+    return;
+  }
+  file.contents = new Buffer(JSON.stringify(file.json, null, 2) + '\n');
+  next(null, file);
+}
+
+function isNull(file, next) {
+  if (typeof file.isNull !== 'function') {
+    file.isNull = function() {
+      return this.contents === null;
+    };
+  }
+  next(null, file);
+}
+
+function isBinary(file, next) {
+  file.isBinary = function() {
+    if (this.hasOwnProperty('_isBinary')) {
+      return this._isBinary;
+    }
+
+    if (this.isNull() || this.isStream() || this.isDirectory()) {
```

Comment:
Since `isStream` and `isDirectory` are used here, there should probably be middleware similar to `isNull` that add those methods to the `file` if they don't exist.